controllers/complianceeventsapi/server.go

```diff
@@ -0,0 +1,334 @@
+
+	return nil
+}
+
+// Stop stops the http server. If it is not currently running, it has no effect.
+func (s *databaseAPIServer) Stop() {
+	s.Lock.Lock()
+	defer s.Lock.Unlock()
+
+	if !s.isRunning {
+		return
+	}
+
+	if err := s.server.Shutdown(context.TODO()); err != nil {
+		log.Error(err, "Error stopping compliance events api server")
+	}
+
+	s.isRunning = false
+}
+
+func postComplianceEvent(goquDB *goqu.Database) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			log.Error(err, "error reading")
+			http.Error(w, "Error reading request body", http.StatusBadRequest)
+
+			return
+		}
+
+		reqEvent := &ComplianceEvent{}
+
+		if err := json.Unmarshal(body, reqEvent); err != nil {
+			log.Error(err, "error unmarshalling")
+			http.Error(w, "Incorrectly formatted request body", http.StatusBadRequest)
+
+			return
+		}
+
+		if err := reqEvent.Validate(); err != nil {
+			log.Error(err, "error validating")
+			http.Error(w, "Invalid request body", http.StatusBadRequest)
+
+			return
+		}
+
+		clusterFK, err := getClusterForeignKey(r.Context(), goquDB, reqEvent.Cluster)
+		if err != nil {
+			log.Error(err, "error getting cluster foreign key")
+			http.Error(w, "Internal Error", http.StatusInternalServerError)
+
+			return
+		}
+
+		reqEvent.Event.ClusterID = clusterFK
+
+		if reqEvent.ParentPolicy != nil {
+			pfk, err := getParentPolicyForeignKey(r.Context(), goquDB, *reqEvent.ParentPolicy)
+			if err != nil {
+				log.Error(err, "error getting parent policy foreign key")
+				http.Error(w, "Internal Error", http.StatusInternalServerError)
+
+				return
+			}
+
+			reqEvent.Policy.ParentPolicyID = &pfk
+		}
+
+		policyFK, err := getPolicyForeignKey(r.Context(), goquDB, reqEvent.Policy)
+		if err != nil {
+			log.Error(err, "error getting policy foreign key")
+			http.Error(w, "Internal Error", http.StatusInternalServerError)
+
+			return
+		}
+
+		reqEvent.Event.PolicyID = policyFK
+
+		insert := goquDB.Insert("compliance_events").Rows(reqEvent.Event).Executor()
+
+		_, err = insert.Exec()
+		if err != nil {
+			log.Error(err, "error inserting compliance event")
+			http.Error(w, "Internal Error", http.StatusInternalServerError)
+		}
+	}
+}
+
+func getClusterForeignKey(ctx context.Context, goquDB *goqu.Database, cluster Cluster) (int, error) {
+	// Check cache
+	key, ok := clusterKeyCache.Load(cluster.ClusterID)
+	if ok {
+		return key.(int), nil
+	}
+
+	foundCluster := new(Cluster)
+
+	found, err := goquDB.From("clusters").
+		Where(goqu.Ex{"cluster_id": cluster.ClusterID}).
+		ScanStructContext(ctx, foundCluster)
+	if err != nil {
+		return 0, err
+	}
+
+	// If the row already exists
+	if found {
+		clusterKeyCache.Store(cluster.ClusterID, foundCluster.KeyID)
+
+		return foundCluster.KeyID, nil
+	}
+
+	// Otherwise, create a new row in the table
+	insert := goquDB.Insert("clusters").Returning("id").Rows(cluster).Executor()
```

Comment:
How is this guarding against SQL injection?

Reply:
My understanding is that the `.Rows(cluster)` part handles any required escaping to prevent injections. Do you think it's missing something?

Reply:
The only info I could find of this online is this:
https://github.com/doug-martin/goqu/issues/279#issuecomment-1367819827

I haven't looked at the code though.